File: javalang/graph/LC_51_NQueens.java
package javalang.graph;

import java.util.*;

public class LC_51_NQueens {
    int n;
    char[][] board;
    boolean[] cols;
    boolean[] diag1;
    boolean[] diag2;
    List<List<String>> result = new ArrayList<>();

    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        board = new char[n][n];
        for (char[] row : board) {
            Arrays.fill(row, '.');
        }

        cols = new boolean[n];
        diag1 = new boolean[2 * n];
        diag2 = new boolean[2 * n];

        dfs(0);
        return result;
    }

    private void dfs(int row) {
        if (row == n) {
            List<String> one = new ArrayList<>();
            for (char[] r : board) {
                one.add(new String(r));
            }
            result.add(one);
            return;
        }

        for (int col = 0; col < n; col++) {
            int d1 = row + col;
            int d2 = row - col + n - 1; // 음수를 피하기 위해 n-1 을 해줘야 한다

            if (cols[col] || diag1[d1] || diag2[d2]) continue;
            // place
            board[row][col] = 'Q';
            cols[col] = diag1[d1] = diag2[d2] = true;

            dfs(row + 1);
            // undo
            board[row][col] = '.';
            cols[col] = diag1[d1] = diag2[d2] = false;
        }
    }
}

File: python/graph/LC_51_NQueens.py
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []

        # 상태 배열
        cols = [False] * n
        diag1 = [False] * (2 * n)  # row + col
        diag2 = [False] * (2 * n)  # row - col + n - 1

        # 결과 표현용 보드
        board = [['.'] * n for _ in range(n)]

        def dfs(row: int):
            # 종료 조건
            if row == n:
                result.append([''.join(r) for r in board])
                return

            for col in range(n):
                d1 = row + col
                d2 = row - col + n - 1

                if cols[col] or diag1[d1] or diag2[d2]:
                    continue

                # place
                board[row][col] = 'Q'
                cols[col] = diag1[d1] = diag2[d2] = True

                dfs(row + 1)

                # undo
                board[row][col] = '.'
                cols[col] = diag1[d1] = diag2[d2] = False

        dfs(0)
        return result